### ISSUE TITLE
HAProxy forwards malformed empty Content-Length headers, in violation of RFC 9110

### ISSUE THREAD

**Issue by kenballus** [haproxy/haproxy#2237]
### Detailed Description of the Problem

RFC 9110 defines the following ABNF rule for the acceptable values in a `Content-Length` header:
```
Content-Length = 1*DIGIT
```
(i.e. 1 or more ASCII digits, so empty values are invalid)

A few paragraphs later, it says this:
> Likewise, a sender MUST NOT forward a message with a Content-Length header field value that does not match the ABNF above, with one exception: a recipient of a Content-Length header field value consisting of the same decimal value repeated as a comma-separated list (e.g, "Content-Length: 42, 42") MAY either reject the message as invalid or replace that invalid field value with a single instance of the decimal value, since this likely indicates that a duplicate was generated or combined by an upstream message processor.

Thus, HAProxy is obligated to reject (or at least, not forward) requests that do not match that ABNF rule.

However, HAProxy currently forwards empty `Content-Length` headers as-is, which is a violation of the RFC.

### Expected Behavior

HAProxy should reject requests with empty `Content-Length` headers.

### Steps to Reproduce the Behavior

1. Build and start HAProxy master. I am using the config below (with the placeholder filled in), but any config should do.
2. Send it the following request:
```
GET / HTTP/1.1\r\n
host: whatever\r\n
content-length: \r\n
\r\n
```
3. Observe that it forwards that request verbatim to your backend, without correcting or rejecting the invalid CL header.

### Do you have an idea how to solve the issue?

Reject empty CL headers.

### What is your configuration?

```haproxy
global
    maxconn 4096

defaults
    mode http
    option http-keep-alive
    timeout client 1s
    timeout connect 1s
    timeout server 1s
    timeout http-request 1s
    http-reuse always

frontend the_frontend
    bind 0.0.0.0:80
    default_backend the_backend

backend the_backend
   server server1 PROXY_BACKEND_PLACEHOLDER:80
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.9-dev2-ef2b15-31 2023/08/02 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 6.4.3-arch1-2 #1 SMP PREEMPT_DYNAMIC Sat, 15 Jul 2023 19:25:49 +0000 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_ZLIB=1 USE_PCRE=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 -BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY -LUA -MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN -SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=32).
Built with OpenSSL version : OpenSSL 3.1.1 30 May 2023
Running on OpenSSL version : OpenSSL 3.1.1 30 May 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with network namespace support.
Built with zlib version : 1.2.13
Running on zlib version : 1.2.13
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.45 2021-06-15
Running on PCRE version : 8.45 2021-06-15
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by kenballus** [comment]
Update: This can be used to get HAProxy to forward requests containing two `Content-Length` headers. If you send this request to HAProxy, it will forward it without correction or rejection:
```
GET / HTTP/1.1\r\nhost: a\r\ncontent-length: \r\ncontent-length: 1\r\n\r\nZ
```

This has obvious request-smuggling implications, and should be fixed ASAP.

**Comment by Tristan971** [comment]
@kenballus and it didn’t occur to you that this isn’t the right place for this comment?

**Comment by kenballus** [comment]
> @kenballus and it didn’t occur to you that this isn’t the right place for this comment?

Sorry; that was a real lapse of judgement. I deleted the comment, and will let the appropriate people know through the official channels.

**Comment by Tristan971** [comment]
Thanks.

fyi pinged @wtarreau about it, though still do send your example to the security address so they can see for themselves. Maybe the example you have is the most one can do with it, or maybe there’s much more to it (in which case it does indeed warrant attention promptly)

**Comment by wtarreau** [comment]
Ben already contacted me. I'll have a look, thanks!

**Comment by wtarreau** [comment]
Now fixed for h1/h2/h3 and with extra regtests (commit 6492f1f). We wanted to produce maintenance releases this end of week, so this will come with them as well. Thanks Ben!

**Comment by rfrohl** [comment]
This issue got CVE-2023-40225 assigned.

**Comment by wtarreau** [comment]
Thank you Robert!

**Comment by robert-scheck** [comment]
I know that HAProxy 1.8 is unmaintained, but could you still comment whether the CVE is applicable for it, please?

**Comment by kenballus** [comment]
I have tested it on haproxy 1.3, and that is also vulnerable. This suggests that 1.8 very likely is too.

**Comment by wtarreau** [comment]
I think it is as well. And in fact older versions work more like a pipe in that they will not transform contents. For example 1.3 will simply pass whatever it receives, a bit like a router... Once we're done with the backport in 2.0 I think the patch will be usable in 1.8 if you really need it.

However, please be aware that it is a problem if a server behind is itself vulnerable to multiple different content-length headers. I don't know which ones are vulnerable to this, but I think that if you're using a serious server like Apache or Nginx behind haproxy you should be safe (Ben might be able to confirm).

As a general rule, you should keep in mind that you shouldn't rely on unmaintained software to secure sensitive components. That can probably be OK if your haproxy doesn't have filtering rules or if the server behind is just a static server, though.

**Comment by kenballus** [comment]
Current versions of nginx and apache are not vulnerable to this. Some smaller web servers (think Python/Ruby ecosystems) are vulnerable. You should test whether your backend server(s) accept requests with 2 content-length headers, of which the first is empty. If it just rejects such requests, then you're probably okay.

**Comment by firexinghe** [comment]
@wtarreau Hello, is 2273176 also a fix patch for CVE-2023-40225?

**Comment by wtarreau** [comment]
No, it's not, even though it was reported by the same person. It's a preventive hardening measure to protect certain poorly written servers against incorrect parsing of leading zeroes. Both of us are aware of the fact that some old non-compliant servers used to suffer from this problem and the protection was easy and inexpensive to add, so it was worth adding it.

**Comment by rodrigoffzz** [comment]
> I think it is as well. And in fact older versions work more like a pipe in that they will not transform contents. For example 1.3 will simply pass whatever it receives, a bit like a router... Once we're done with the backport in 2.0 I think the patch will be usable in 1.8 if you really need it.

Hi,
I'm working on a downstream backport for Ubuntu and I noticed that the patch for 2.4 is cleanly applicable for 2.0 (with only one change in a comment line), if that is a correct understanding, do you think you could, please, give an idea on what would be the fix for 1.8, where `*_parse_cont_len_header` methods did not exist yet, if feasible?

Going a bit further in the past, before 1.8, where even `*_headers_to_hdr_list` did not exist yet, I couldn't find where we could place a possible fix. As already stated, older versions are routing whatever it receives, so it would probably don't get a proper fix for this kind of issue since it was not checking it at all... do you think this is a valid statement?

thank you very much !

**Comment by wtarreau** [comment]
Note, for 2.0 there are two HTTP stacks, the legacy and the HTX one. It was a transition version. The part from 2.4 that applies well to 2.0 is only for HTX, we still need to do the backport for the legacy one. I tend to think that once we have a complete one for 2.0 it will be OK to backport it as-is to 1.8.

Once I'm done with the backports, I'll have a look at 1.8.

**Comment by git001** [comment]
> I know that HAProxy 1.8 is unmaintained, but could you still comment whether the CVE is applicable for it, please?

Isn't that the perfect reason to update to a more recent and maintained version?

**Comment by git001** [comment]
> Note, for 2.0 there are two HTTP stacks, the legacy and the HTX one. It was a transition version. The part from 2.4 that applies well to 2.0 is only for HTX, we still need to do the backport for the legacy one. I tend to think that once we have a complete one for 2.0 it will be OK to backport it as-is to 1.8.
> 
> Once I'm done with the backports, I'll have a look at 1.8.

But isn't 1.8 EOL? Just for my curiosity why invest some time to backport and release an new version for a EOL version?

**Comment by wtarreau** [comment]
Yes it's dead. But you know how it is, Aleks. First, there are LTS distros which ship with it and will need the fix. Second, users never upgrade in emergency, that's the best way to face an error due to a rushed upgrade when they would otherwise not have been vulnerable :-)  I'll have a look in best effort mode, that's fine. It shouldn't be different from the 2.0 fix anyway.

**Comment by wtarreau** [comment]
@rodrigoffzz the great news is that the problem only affects HTX mode, thus 1.9+. The legacy mode in 2.0 and before has the correct check for empty value which immediately returns a 400 bad request. This check was introduced from the beginning when we started to support keep-alive, in 1.4-dev5 with commit 32b47f42a0, which explains why @kenballus found 1.3 to be affected. In fact 1.3 doesn't support keep-alive so it will pass anything as-is and will not even parse content-length. In that sense it's not affected since the right way to use it is http-close (good old times where keep-alive didn't exist :-)).

So you're fine with 1.8. However, there is still the fact that we'll pass leading zeroes as-is. This could only be addressed in 2.0+ (with HTX), it's really not feasible in older versions due to the way values are indexed (spaces are already skipped in the index). Regardless, as Ben indicated it's not a bug and what we pass is perfectly valid, it's just that we know some rare poor-quality servers can be fooled while we could have had the opportunity to clean what we pass to them. Knowing that from 2.0 and above (hence all supported versions) we can pass them something clean is already OK IMHO.

**Comment by rodrigoffzz** [comment]
> @rodrigoffzz the great news is that the problem only affects HTX mode, thus 1.9+. The legacy mode in 2.0 and before has the correct check for empty value which immediately returns a 400 bad request. This check was introduced from the beginning when we started to support keep-alive, in 1.4-dev5 with commit [32b47f4](https://github.com/haproxy/haproxy/commit/32b47f42a0d131d3889c9292a2bb06388a752969), which explains why @kenballus found 1.3 to be affected. In fact 1.3 doesn't support keep-alive so it will pass anything as-is and will not even parse content-length. In that sense it's not affected since the right way to use it is http-close (good old times where keep-alive didn't exist :-)).
> 
thank you very much for getting back with that many details, it is indeed good news :)
> So you're fine with 1.8. However, there is still the fact that we'll pass leading zeroes as-is. This could only be addressed in 2.0+ (with HTX), it's really not feasible in older versions due to the way values are indexed (spaces are already skipped in the index). Regardless, as Ben indicated it's not a bug and what we pass is perfectly valid, it's just that we know some rare poor-quality servers can be fooled while we could have had the opportunity to clean what we pass to them. Knowing that from 2.0 and above (hence all supported versions) we can pass them something clean is already OK IMHO.
>
got it, it does make sense.

thanks again!

**Comment by kenballus** [comment]
> So you're fine with 1.8. However, there is still the fact that we'll pass leading zeroes as-is. This could only be addressed in 2.0+ (with HTX), it's really not feasible in older versions due to the way values are indexed (spaces are already skipped in the index). Regardless, as Ben indicated it's not a bug and what we pass is perfectly valid,

Agreed.

> ...it's just that we know some rare poor-quality servers can be fooled while we could have had the opportunity to clean what we pass to them.

This isn't as rare as you might think. If you're worried about this bug, you should check whether your server times out when you send it the following request:
```
GET / HTTP/1.1\r\n
Host: whatever\r\n
Content-Length: 010\r\n
\r\n
01234567
```
If it times out, then your server is unaffected. If it doesn't time out, then you should consider updating.

**Comment by wtarreau** [comment]
This is an excellent idea indeed! Thanks!
